sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -3416,6 +3416,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val DECORRELATE_EXISTS_AND_IN_SUBQUERIES =
+    buildConf("spark.sql.optimizer.decorrelateExistsIn.enabled")
+      .internal()
+      .doc("Decorrelate EXISTS and IN subqueries.")
+      .version("4.0.0")
+      .booleanConf
+      .createWithDefault(true)
```

Comment:
Will this have any query result changes if we enable this by default?

Reply:
some queries were not supported and are supported with this flag, that's the only result change.

Reply:
I think there might be some correctness fixes here for COUNT bug in IN/EXISTS too, enabled by moving those to DecorrelateInnerQuery, right? I know not every query is fixed but there are some that are fixed, such as the example query you have added to one of your tests `select * from t1 where exists (select count(*) from t2 where t2.c1 = t1.c1);`

Reply:
changed the flag to reflect that there is some legacy behavior. Added tests for that behavior.